### ISSUE TITLE
cu3 simulated matrix false on unitary simulator

### ISSUE THREAD

**Issue by nelimee** [Qiskit/qiskit-terra#546]
**Informations**

- **QISKit (Python SDK) version**: 0.5.3
- **Operating system**: Fedora 26
- **Python version**: 3.6.5

**What is the current behavior?**

The `cu3` gate simulated with the `local_unitary_simulator` does not produce the expected matrix for a `cu3` gate.

**Steps to reproduce the problem**

Execute the script:
```py
import qiskit
from qiskit import available_backends, get_backend,\
    execute, QuantumRegister, ClassicalRegister, QuantumCircuit
import numpy as np

def round_to_zero(vec, tol=2e-15):
    vec.real[abs(vec.real) < tol] = 0.0
    vec.imag[abs(vec.imag) < tol] = 0.0
    return vec

def controlled(U):
    n = U.shape[0]
    return np.vstack((np.hstack((np.identity(n), np.zeros((n,n)))),
                      np.hstack((np.zeros((n,n)), U))))

def U(theta, phi, lam):
    return np.array([
        [np.cos(theta/2),
         -np.exp(1.j*lam)*np.sin(theta/2)],
        [np.exp(1.j*phi)*np.sin(theta/2),
         np.exp(1.j*(phi+lam))*np.cos(theta/2)]
    ])

Q_SPECS = {
    "name": "TestCU3",
    "circuits": [
        {
            "name": "TestCU3",
            "quantum_registers": [
                {
                    "name": "ctrl",
                    "size": 1
                },
                {
                    "name": "qb",
                    "size": 1
                },
            ],
            "classical_registers": []
        }
    ],
}
Q_program = qiskit.QuantumProgram(specs=Q_SPECS)

circuit = Q_program.get_circuit("TestCU3")
qb = Q_program.get_quantum_register("qb")
ctrl = Q_program.get_quantum_register("ctrl")

theta, phi, lam = np.pi, np.pi/2, 3*np.pi/2

circuit.cu3(theta, phi, lam, ctrl[0], qb[0])

unitary_sim = get_backend('local_unitary_simulator')
res = execute([circuit], unitary_sim).result()
unitary = round_to_zero(res.get_unitary())
wanted_unitary = round_to_zero(controlled(U(theta, phi, lam)))

print("Unitary matrix from simulator:\n", unitary, sep='')
print("Unitary matrix from theory:\n", wanted_unitary, sep='')
print("Unitary matrices are the same:", np.allclose(unitary, wanted_unitary))
```

Output on my machine:
```
Unitary matrix from simulator:
[[1.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.-1.j]
 [0.+0.j 0.+0.j 1.+0.j 0.+0.j]
 [0.+0.j 0.-1.j 0.+0.j 0.+0.j]]
Unitary matrix from theory:
[[1.+0.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 1.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j 0.+1.j]
 [0.+0.j 0.+0.j 0.+1.j 0.+0.j]]
Unitary matrices are the same: False
```

**What is the expected behavior?**

The two printed matrices should be equal.

**Suggested solutions**

The problem might be in:

1. The unitary simulator.
2. The definition of the `cu3` gate in `qiskit.extensions.standard.header`. 

I don't have any solution for the moment.

**Comment by chriseclectic** [comment]
It appears there that the definition of `cu3` in the standard header is actually implementing the a controlled-U where `U = exp( -1j * (phi + lam)/2) * U3(theta, phi, lam)`.  So there is a local phase which wouldn't matter for a single `u3` (in this case it would be a global phase), but is important when implementing the controlled gate.

(Less important but you swapped the target and control qubits between the two cases,  the "standard" cnot in qiskit is `CX(0, 1) = [[1, 0, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0], [0, 1, 0, 0]]` since qubit 0 is to the right of the tensor product).

**Comment by jaygambetta** [comment]
This is in the tutorial and i have not decided what we should do.

https://github.com/QISKit/qiskit-tutorial/blob/master/reference/tools/quantum_gates_and_linear_algebra.ipynb

i am leaning towards deleting this gate and making a new one that we call cU which is a four parameter gate. I would also like to change the name of cU1(theta) to cphase(theta)

**Comment by rraymondhp** [comment]
I wrote a note emphasizing the importance of global phase in control-U and its difference with just U. But that part is now missing.

**Comment by jaygambetta** [comment]
cU3 is ambiguous and i agree we need to make it clearer. A real cU should have 4 variables not 3. The tensor order has been described in many places (see the tutorials) and it will be clearer in the documentation when we redo it.

**Comment by chriseclectic** [comment]
Closed by #2755
